### ISSUE TITLE
P2542 R7 views::concat

### ISSUE THREAD

**Issue by wg21bot** [cplusplus/papers#1204]
[P2542R0](https://wg21.link/p2542r0) views::concat (Hui Xie, S. Levent Yilmaz)

**Comment by inbal2l** [comment]
Was seen by SG9 (Ranges) on 2022-03-14 ([Full Minutes](https://wiki.edg.com/bin/view/Wg21telecons2022/P2542#SG9-2022-03-14))

### Polls

No polls were taken.

### Summary

Paper is in good shape. Needs revision addresses the topics of: (Whether to) Fix 4.5 Common Range, usage of common_reference_with.
We will take a poll on the next revision.

**Comment by wg21bot** [comment]
[P2542R1](https://wg21.link/p2542r1) views::concat (Hui Xie, S. Levent Yilmaz)

**Comment by inbal2l** [comment]
R1 was seen by SG9 (Ranges) on 2022-04-11 ([Full Minutes](https://wiki.edg.com/bin/view/Wg21telecons2022/P2542#SG9-2022-04-11))

### Polls

No polls were taken.

### Summary

The issue of usage of common_reference_with with prvalue, xvalues, and the ownership was brought up. 
Should be addressed by authors in the mailing before moving forward with the paper. 
Will be seen at the next SG9 meeting (on 2022-05-09)

**Comment by inbal2l** [comment]
R2 was seen by SG9 (Ranges) on 2022-05-09 ([Full Minutes](https://wiki.edg.com/bin/view/Wg21telecons2022/P2542#SG9-2022-05-09))

### Polls


__POLL: Forward “P2542R2: views::concat” to LEWG for C++26.__

|SF|F|N|A|SA|
|-|-|-|-|-|
| 5  |  1 | 0  | 0  |  0 |

__Author’s position:__ 2xSF
__Attendance:__ 10
__Outcome:__ Strong consensus in favor

### Summary

Revision addresses the topics of usage of common_reference_with.
Motivation, design, feature test macro, wording and implementation experiance all exists. There was a suggestion (by the author) to add example in the wording.
R2 of the paper was forwarded to LEWG for C++26.

**Comment by wg21bot** [comment]
[P2542R2](https://wg21.link/p2542r2) views::concat (Hui Xie, S. Levent Yilmaz)

**Comment by brycelelbach** [comment]
## [2022-07 Library Evolution Electronic Poll Outcomes](https://wg21.link/P2611)

__Poll 2.5: Send [P2542R2] views::concat to Library Working Group for C++26, classified as an addition ([P0592R4] bucket 3 item).__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|10|7|0|0|0|

__Outcome: Unanimous consensus in favor.__

**Comment by inbal2l** [comment]
The topic of promoting this paper for C++23 (as the feature providing support for "ranges" composition) was brought up in SG9 on 2022-10-10  ([Full Minutes](https://wiki.edg.com/bin/view/Wg21telecons2022/P2542#SG9-2022-10-10)).

### Polls

__POLL: We see the suggested utility “P2542R2: views::concat” important enough to be suggested for C++23 (as an NB comment)__

|SF|F|N|A|SA|
|-|-|-|-|-|
|  1 | 2  | 0  | 2  |  1  |

__Attendance:__ 10

__Author’s position:__ SFx1

__Outcome:__ No consensus

### Summary

SG9 does not see the urgency of the feature and believes it can be delivered in C++26 as planned. We suspect the rest of WG21 will see it similarly. Authors are of course welcome to try and promote this as an NB comment for C++23.

**Comment by JeffGarland** [comment]
@inbal2l We didn't get an update so library will not see this in Issaquah

**Comment by inbal2l** [comment]
SG9 (Ranges) reviewed [P2542R3](https://isocpp.org/files/papers/P2542R3.html) during the Varna meeting on 2023-06-12 ([Full Minutes](https://wiki.edg.com/bin/view/Wg21varna/P2542)).

### Polls

__POLL__: Forward “P2542R3: views::concat” to LWG for C++26 (with the noexcept policy defined in the “policy” discussion).

|SF|F|N|A|SA|
|-|-|-|-|-|
|  3 |2  | 1 |0  | 0 |

__Attendance:__ 8

__# of Authors:__ 1

__Author Position:__ SF

__Outcome:__ Strong consensus in favor

### Summary

SG9 reviewed the updates to the paper: 

- Relaxed random_access_range constraints. 
- Redesigned iter_swap to solve the issues (concat(range of string&&, range of string) is broken)
- The topic of "noexcept" also came up (as there's a pre-condition that the object is not dereferenceable). As of now, we don't have a "noexcept" policy, so SG9 would like to get this input from LEWG (this should probably not block this paper, as having such a policy will have a wide-scale effect).

The authors will make sure all the other wording changes requested by LWG have been addressed before going back to them.

**Comment by hewillk** [comment]
Hum, I'm not sure if I can comment here, if not I'll delete my comment and apologize for the recklessness.
There is a potential issue here, `std::variant<Args...>` can not accept `std::variant<Brgs...>` like range-v3's `variant`, so `iterator(iterator<!Const> i)` currently won't work, and I don't think there's an easy way to make it work.

**Comment by jensmaurer** [comment]
@hewillk, this issue tracker is exclusively for administrative purposes. Feel free to send any technical comments directly to the authors or to a suitable committee reflector.

**Comment by wg21bot** [comment]
[P2542R4](https://wg21.link/p2542r4) views::concat (Hui Xie, S. Levent Yilmaz)

**Comment by wg21bot** [comment]
[P2542R5](https://wg21.link/p2542r5) views::concat (Hui Xie, S. Levent Yilmaz)

**Comment by inbal2l** [comment]
SG9 (Ranges) reviewed [P2542R4](https://wg21.link/P2542R4) during September 11th telecon 2023-09-11 ([Full Minutes](https://wiki.edg.com/bin/view/Wg21telecons2023/P2542)).

### Polls

__POLL: We consider the use-case below as "use after move", and recommend that views::concat would not handle this explicitly:__
```cpp
auto bad_mix = concat(v1 | views::as_rvalue, v2 | copy);
std::cout << "ISSUE\n";
std::cout << " 1st deref: " << *bad_mix.begin() << "\n"; // prints "a"
std::cout << " 2nd deref: " << *bad_mix.begin() << "\n"; // prints ""
```
|SF|F|N|A|SA|
|-|-|-|-|-|
|3|3|0|0|0|

__Authors:__ 2xSF
__Attendance:__ 9
__Outcome:__ Strong consensus in favor

__POLL: SG9 would recommend going back to constraints in a similar form suggested in P5 of R0 of "views::concat" (https://isocpp.org/files/papers/P2542R0.html#concat-view-range.concat) Which means phrase the requirements to support the use case for "non-common and random access + sized" range.__

|SF|F|N|A|SA|
|-|-|-|-|-|
|2|0|2|1|0|

__Authors:__ 1xA 1xN
__Attendance:__ 9
__Outcome:__ No concensus

Daisy: Please prepare examples of user code that will be affected by this change.

### Summary

SG9 recommended the following:

- Apply the requirements: reject the type by an extra constraint that if (a) the resulting concat_view's reference_t is a prvalue (marked T) and (b) any one of the r... has reference_t T& or T&& (keeping the `views::concat` behavior similar to existing views).
- There was no consensus to revert the requirements to the one that appears in `cartisian_product`. 
 
However, for the second part, to prepare for LEWG discussion, we asked the authors to come up with use cases that will be blocked by the less-fine-grained requirement (only require "common_range" instead of "non-common and random access + sized" range).

The paper was returned to LEWG with the recommendations above.

**Comment by ben-craig** [comment]
## P2542R5: views::concat

[P2542R5](https://wg21.link/P2542R5): views::concat

[2023-09-19 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21telecons2023/P2542#Library-Evolution-2023-09-19)

__Champion:__ Levent Yilmaz

__Chair:__ Ben Craig

__Minute Taker:__ Guy Davidson

__Start:__ 2023-09-19

### Summary

__POLL: Change concat-is-random-access and concat-is-bidi requirements from (common_range or sized_range) to only common_range __

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|2|8|1|1|0|

__Attendance:__ 22

__# of Authors:__ 1

__Author Position:__ WF

__Outcome:__ Consensus

WA: Rather have complexity in implementation rather than user code for corner cases

### Next Steps

Postpone views::concat R5 LEWG electronic poll, and put R6 in the next LEWG electronic poll

**Comment by wg21bot** [comment]
[P2542R6](https://wg21.link/p2542r6) views::concat (Hui Xie)

**Comment by inbal2l** [comment]
[P2542R7](https://wg21.link/P2542R7) was added to LEWG 2023 December electronic poll paper ([P3053R0](https://wg21.link/P3053R0))

**Comment by inbal2l** [comment]
[P2542R7](https://wg21.link/p2542r7) was forwarded to LWG on LEWG 2023 December electronic poll ([P3054R0](https://wg21.link/P3054R0))

**Comment by JeffGarland** [comment]
Poll: put P2542r8 into C++26?
LWG reviewed and approved in Tokyo.

```
|F|A|N|
|7|0|0|
```

**Comment by wg21bot** [comment]
[P2542R8](https://wg21.link/p2542r8) views::concat (Hui Xie, S. Levent Yilmaz)
